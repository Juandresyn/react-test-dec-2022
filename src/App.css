#root {
  width: 1100px;
  margin: 0 auto;
  padding: 2rem;
  text-align: center;
  --nav-bar-width: 230px;
}

.logo {
  height: 6em;
  padding: 1.5em;
  will-change: filter;
}
.logo:hover {
  filter: drop-shadow(0 0 2em #646cffaa);
}
.logo.react:hover {
  filter: drop-shadow(0 0 2em #61dafbaa);
}

@media (prefers-reduced-motion: no-preference) {
  a:nth-of-type(2) .logo {
    animation: logo-spin infinite 20s linear;
  }
}

.card {
  padding: 2em;
}

.main {
  display: flex;
}

.toolbar {
  width: var(--nav-bar-width);
  text-align: left;
  padding-right: 30px;
  margin-top: 120px;
}

.menu {
  padding: 0;
  margin: 0;
  list-style: none;
  text-transform: uppercase;
}

.menu li {
  padding: 10px 0;
}

.menu a {
  font-weight: bold;
}

.content {
  width: calc(100% - var(--nav-bar-width));
}