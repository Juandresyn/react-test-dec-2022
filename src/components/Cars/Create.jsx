import { useState } from 'react';
import PropTypes from 'prop-types';

function CreateReservation({ submit }) {
  const [id, setId] = useState(null);
  const [maker, setMaker] = useState(null);
  const [model, setModel] = useState(null);
  const [ref, setRef] = useState(null);
  const [color, setColor] = useState(null);
  const [milage, setMilage] = useState(null);

  /**
   * Default object representation
   */
  const responseBody = {
    id: '',
    maker: '',
    model: 0,
    ref: '',
    color: '',
    milage: 0,
  };

  /**
   * Handle form submit
   */
  const handleSubmit = (event) => {
    event.preventDefault();

    responseBody.id = id;
    responseBody.maker = maker;
    responseBody.model = model;
    responseBody.ref = ref;
    responseBody.color = color;
    responseBody.milage = milage;

    submit(responseBody);
  };

  const inputChangeHandler = (setFunction, event) => setFunction(event.target.value);

  return (
    <form className="form CreateReservation" onSubmit={handleSubmit}>
      <div className="form__wrapper">
        <input role="textbox" type="text" placeholder="License" name="id" onChange={(e) => inputChangeHandler(setId, e)} />
        <input role="textbox" type="text" placeholder="Maker" name="maker" onChange={(e) => inputChangeHandler(setMaker, e)} />
        <input role="textbox" type="text" placeholder="Model" name="Model" onChange={(e) => inputChangeHandler(setModel, e)} />
        <input role="textbox" type="text" placeholder="Ref" name="Ref" onChange={(e) => inputChangeHandler(setRef, e)} />
        <input role="textbox" type="text" placeholder="Color" name="Color" onChange={(e) => inputChangeHandler(setColor, e)} />
        <input role="textbox" type="text" placeholder="Milage" name="Milage" onChange={(e) => inputChangeHandler(setMilage, e)} />
      </div>
      <button role="button">Create Car</button>
    </form>
  );
}

CreateReservation.propTypes = {
  submit: PropTypes.func.isRequired,
};

export default CreateReservation;
