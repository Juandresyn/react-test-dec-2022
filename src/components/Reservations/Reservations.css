.reservations__cta {
    width: 100%;
    margin-bottom: 30px;
}