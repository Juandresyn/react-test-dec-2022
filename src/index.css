:root {
  font-family: Inter, Avenir, Helvetica, Arial, sans-serif;
  font-size: 16px;
  line-height: 24px;
  font-weight: 400;

  color-scheme: light dark;
  color: rgba(255, 255, 255, 0.87);
  background-color: #242424;

  font-synthesis: none;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-text-size-adjust: 100%;
}

a {
  font-weight: 500;
  color: #646cff;
  text-decoration: inherit;
}
a:hover {
  color: #535bf2;
}

body {
  margin: 0;
  display: flex;
  min-width: 320px;
  min-height: 100vh;
}

h1 {
  font-size: 3.2em;
  line-height: 1.1;
}

button {
  border-radius: 8px;
  border: 1px solid transparent;
  padding: 0.6em 1.2em;
  font-size: 1em;
  font-weight: 500;
  font-family: inherit;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.25s;
}
button:hover {
  border-color: #646cff;
}
button:focus,
button:focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

.button--wide {
  width: 100%;
  margin-bottom: 30px;
}

.form {
  display: flex;
  flex-direction: column;
  margin-bottom: 35px;
}

.form__wrapper {
  display: flex;
  flex-direction: column;
}

.form__group {
  display: flex;
  justify-content: space-between;
}

.form__group > * {
  width: 49% !important;
}

.form__wrapper input,
.form__wrapper select,
.form__wrapper textarea {
  width: 100%;
  padding: 15px 10px;
  border: 1px solid #cdcdcd;
  border-radius: 3px;
  margin-bottom: 14px;
  box-sizing: border-box;
}

.title__main {
  text-align: left;
}

@media (prefers-color-scheme: light) {
  :root {
    color: #213547;
    background-color: #ffffff;
  }
  a:hover {
    color: #747bff;
  }
  button {
    background-color: #f9f9f9;
  }
}
