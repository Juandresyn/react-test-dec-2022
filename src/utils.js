export const transformDate = (date, showTime = true) => {
    const year = date.getFullYear();
    const month = date.getMonth() + 1;
    const day = date.getDate();
    const hour = date.getHours();
    const minutes = date.getMinutes();

    const time = `${hour}:${minutes}:00`;

    return `${year}-${month}-${day} ${showTime ? time : ''}`
};

export const getAge = (dob) => {
    const ageDifMs = Date.now() - new Date(dob).getTime();
    const ageDate = new Date(ageDifMs);

    return Math.abs(ageDate.getUTCFullYear() - 1970);
}

export const users = [
    {
        id: '23456789',
        name: 'Pepito',
        lastName: 'Perez',
        dob: '1973-06-12'
    },
    {
        id: '12345678',
        name: 'Juan',
        lastName: 'Yepes',
        dob: '1992-02-16'
    }
];

export const cars = [
    {
      id: "NOF123",
      maker: 'mercedes-benz',
      model: 2023,
      ref: 'GLC300',
      color: 'blue',
      milage: 31278
    },
    {
      id: "IHJ889",
      maker: 'audi',
      model: 2020,
      ref: 'Q5',
      color: 'white',
      milage: 45123
    },
    {
      id: "UUE654",
      maker: 'mazda',
      model: 2021,
      ref: 'CX-9',
      color: 'silver',
      milage: 55687
    }
];

export const reservations = [
    {
      id: "rid-20221210-000001",
      user: {
        id: '12345678',
        name: 'Juan',
        lastName: 'Yepes',
        age: '30'
      },
      car: {
        id: "NOF123",
        maker: 'mercedes-benz',
        model: 2023,
        ref: 'GLC300',
        color: 'blue',
        milage: 31278
      },
      from: '2022-12-10 10:00:00',
      to: '2022-12-20 11:00:00',
      notes: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed volutpat, nibh nec egestas sodales, justo nulla sodales dolor.'
    },
    {
      id: "rid-20221210-000002",
      user: {
        id: '23456789',
        name: 'Pepito',
        lastName: 'Perez',
        age: '45'
      },
      car: {
        id: "IHJ889",
        maker: 'audi',
        model: 2020,
        ref: 'Q5',
        color: 'white',
        milage: 45123
      },
      from: '2022-12-10 10:30:00',
      to: '2022-12-20 11:10:00',
      notes: ''
    },
];