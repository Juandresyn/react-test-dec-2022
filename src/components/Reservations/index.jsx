import { useState, useCallback, useEffect } from 'react';
import useFetch, { Provider } from 'use-http';
import Table from '../Table';
import Create from './Create';
import { ENDPOINT_BASE_URL, RESERVATIONS_EP } from '../../config/api';

const headers = ['id', 'user', 'car', 'from', 'to'];

const transformRows = (row) => (row.user ? {
  id: row.id,
  user: `${ row.user.name } ${ row.user.lastname }`,
  car: `${ row.car.ref } ${ row.car.model } ${ row.car.color }`,
  from: row.from,
  to: row.to
} : {});

function Reservations() {
  const [showForm, setShowForm] = useState(false);
  const [reservationsList, setReservationsList] = useState([]);

  /**
   * Initialize useFetch
   */
  const { get, post, del, response, loading, error } = useFetch(ENDPOINT_BASE_URL, { data: [] });

  /**
   * Load reservations from endpoint
   */
  const loadInitialReservations = useCallback(async () => {
    const initialReservations = await get(RESERVATIONS_EP);
    if (response.ok) setReservationsList(initialReservations.data)
  }, [get, response]);

  /**
   * componentDidMount
   */
  useEffect(() => { loadInitialReservations() }, [loadInitialReservations]);

  /**
   * send new reservation data to the server for creation
   */
  const postNewReserv = useCallback(async (data) => {
    if (!!data && !data.userId) return;
    const newReserv = await post(RESERVATIONS_EP, data);

    if (response.ok) {
      setReservationsList(items => [newReserv.reservation, ...items]);
      setShowForm(showForm => false);
    }
  }, [post, response, reservationsList])

  /**
   * Ask server to delete an existing reservation.
   */
  const deleteReserv = useCallback(async (id) => {
    await del(`${ RESERVATIONS_EP }/${id}`);

    if (response.ok) {
      loadInitialReservations();
    }
  }, [del, response, reservationsList]);

  /**
   * Handle for when <Create> is submitted
   */
  const handleSubmit = async (data) => {
    postNewReserv(data);
  };

  return (
    <div className="Reservations">
      <h1 className="title__main">Manage Reservations</h1>
      <button role="button" className="button--wide" onClick={() => setShowForm((showForm) => !showForm)}>{ showForm ? 'Hide Form' : 'Create Reservation' }</button>

      { showForm ? <Create submit={ (data) => handleSubmit(data) }></Create> : null }

      { !loading ? `${ reservationsList.length } reservations found` : '' }

      <Table
        headers={ headers }
        items={ reservationsList.map(transformRows) }
        options={({
          remove: deleteReserv
        })}></Table>
    </div>
  )
}

export default Reservations;
