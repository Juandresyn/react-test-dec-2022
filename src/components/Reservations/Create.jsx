import { useState } from 'react';
import DatePicker from 'react-datepicker';
import "react-datepicker/dist/react-datepicker.css";
import { transformDate } from '../../utils';

function CreateReservation({ submit }) {
  const [user, setUser] = useState(null);
  const [car, setCar] = useState(null);
  const [notes, setNotes] = useState(null);
  const [startDate, setStartDate] = useState(null);
  const [endDate, setEndDate] = useState();

  /**
   * Default object representation
   */
  const responseBody = {
    userId: '',
    carId: '',
    from: null,
    to: null,
    notes: ''
  };

  /**
   * Handle form submit
   */
  const handleSubmit = (event) => {
    event.preventDefault();
    responseBody.userId = parseInt(user);
    responseBody.carId = car;
    responseBody.from = transformDate(startDate);
    responseBody.to = transformDate(endDate);
    responseBody.notes = notes;

    submit(responseBody);
  };

  /**
   * Handle form when and input changes its value.
   */
  const inputChangeHandler = (setFunction, event) => setFunction(event.target.value);

  return (
    <form className="form CreateReservation" onSubmit={handleSubmit}>
      <div className="form__wrapper">
          <input type="text" placeholder="User" name="user" onChange={(e)=>inputChangeHandler(setUser, e)} />
          <input type="text" placeholder="Car" name="car" onChange={(e)=>inputChangeHandler(setCar, e)} />
          <div className="form__group">
            <DatePicker
              selected={startDate}
              showTimeSelect
              onChange={(date) => setStartDate(date)}
              minDate={new Date()}
              selectsStart
              showDisabledMonthNavigation
              endDate={endDate}
              placeholderText="From"
            />
            <DatePicker
              selected={endDate}
              showTimeSelect
              onChange={(date) => setEndDate(date)}
              selectsEnd
              startDate={startDate}
              endDate={endDate}
              minDate={startDate}
              placeholderText="To"
            />
          </div>
          <textarea name="notes" id="" cols="30" rows="5" placeholder="Notes" onChange={(e)=>inputChangeHandler(setNotes, e)}></textarea>
      </div>
      <button>Create Reservation</button>
    </form>
  )
}

export default CreateReservation;
