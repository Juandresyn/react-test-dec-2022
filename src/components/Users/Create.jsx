import { useState } from 'react';
import DatePicker from 'react-datepicker';
import "react-datepicker/dist/react-datepicker.css";
import { transformDate } from '../../utils';

function CreateReservation({ submit }) {
  const [id, setId] = useState(null);
  const [name, setName] = useState(null);
  const [lastname, setLastname] = useState(null);
  const [dob, setDob] = useState(null);

  const responseBody = {
    id: '',
    name: '',
    lastname: 0,
    dob: null
  };

  const handleSubmit = (event) => {
    event.preventDefault();

    responseBody.id = id;
    responseBody.name = name;
    responseBody.lastname = lastname;
    responseBody.dob = transformDate(dob, false);

    submit(responseBody);
  };

  const inputChangeHandler = (setFunction, event) => setFunction(event.target.value);

  return (
    <form className="form CreateReservation" onSubmit={handleSubmit}>
      <div className="form__wrapper">
          <input type="number" placeholder="ID" name="id" onChange={(e)=>inputChangeHandler(setId, e)} />
          <input type="text" placeholder="Name" name="maker" onChange={(e)=>inputChangeHandler(setName, e)} />
          <input type="text" placeholder="Last Name" name="Model" onChange={(e)=>inputChangeHandler(setLastname, e)} />
          <DatePicker
              selected={dob}
              onChange={(date) => setDob(date)}
              peekNextMonth
              showMonthDropdown
              showYearDropdown
              dropdownMode="select"
              placeholderText="DoB"
              maxDate={Date.now()}
            />
      </div>
      <button>Create User</button>
    </form>
  )
}

export default CreateReservation;
