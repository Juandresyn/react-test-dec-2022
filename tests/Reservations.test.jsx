import {
  render, screen, waitFor, fireEvent, act,
} from '@testing-library/react';
import { useFetch } from 'use-http';
import Reservations from '../src/components/Reservations';
import Create from '../src/components/Reservations/Create';
import { reservations } from '../src/mockData';
import * as utils from '../src/utils';

jest.mock('use-http');

jest.mock('react-datepicker/dist/react-datepicker.css', () => ({
  __esModule: true,
  default: {
    class1: 'class1',
  },
}));

describe('Test <Reservations /> component', () => {
  useFetch.mockReturnValue({
    get: () => ({ data: reservations }),
    post: () => null,
    del: () => null,
    response: { ok: true },
    loading: false,
  });

  test('Should call endpoint', async () => {
    let Render;

    act(() => {
      Render = render(<Reservations />);
    });

    await waitFor(() => {
      const plainHTML = Render.asFragment();
      expect(plainHTML).toMatchSnapshot();

      const carsCount = screen.getByText(`${reservations.length} reservations found`);
      expect(carsCount).toBeTruthy();
    });
  });

  test('Should open <Create /> component', async () => {
    let Render;

    act(() => {
      Render = render(<Reservations />);
    });

    await waitFor(() => {
      fireEvent.click(Render.container.querySelector('[role="button"]'));

      const button = screen.getByRole('button', {
        name: /hide form/i,
      });

      expect(button).toBeTruthy();
    });
  });
});

describe('Test <Create /> component', () => {
  test('<Create /> component should pass data on submit', async () => {
    const createDataMock = {
      userId: 1,
      carId: '2',
      from: null,
      to: null,
      notes: 'Lorem ipsum dolor sit',
    };
    let Render;
    const handleSubmitMock = jest.fn();

    act(() => {
      Render = render(<Create submit={handleSubmitMock} />);
    });

    await waitFor(() => {
      const inputId = Render.container.querySelector('[name="user"]');
      fireEvent.change(inputId, { target: { value: createDataMock.userId } });

      expect(inputId.value).toBe(createDataMock.userId.toString());

      const inputName = Render.container.querySelector('[name="car"]');
      fireEvent.change(inputName, { target: { value: createDataMock.carId } });

      expect(inputName.value).toBe(createDataMock.carId.toString());

      const inputNotes = Render.container.querySelector('[name="notes"]');
      fireEvent.change(inputNotes, { target: { value: createDataMock.notes } });

      expect(inputNotes.value).toBe(createDataMock.notes);

      // eslint-disable-next-line no-import-assign
      utils.transformDate = jest.fn(() => null);

      fireEvent.click(Render.container.querySelector('[role="button"]'));

      expect(handleSubmitMock).toBeCalledWith(createDataMock);
    });
  });
});
