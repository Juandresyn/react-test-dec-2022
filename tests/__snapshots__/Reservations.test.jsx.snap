// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <Reservations /> component Should call endpoint 1`] = `
<DocumentFragment>
  <div
    class="Reservations"
  >
    <h1
      class="title__main"
    >
      Manage Reservations
    </h1>
    <button
      class="button--wide"
      role="button"
    >
      Create Reservation
    </button>
    0 reservations found
    <table
      class="table"
    >
      <thead>
        <tr>
          <th>
            ID
          </th>
          <th>
            USER
          </th>
          <th>
            CAR
          </th>
          <th>
            FROM
          </th>
          <th>
            TO
          </th>
          <th>
            OPTIONS
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</DocumentFragment>
`;

exports[`Test <Reservations /> component Should call endpoint 2`] = `
<DocumentFragment>
  <div
    class="Reservations"
  >
    <h1
      class="title__main"
    >
      Manage Reservations
    </h1>
    <button
      class="button--wide"
      role="button"
    >
      Create Reservation
    </button>
    2 reservations found
    <table
      class="table"
    >
      <thead>
        <tr>
          <th>
            ID
          </th>
          <th>
            USER
          </th>
          <th>
            CAR
          </th>
          <th>
            FROM
          </th>
          <th>
            TO
          </th>
          <th>
            OPTIONS
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          id="rid-20221210-000001"
        >
          <td>
            rid-20221210-000001
          </td>
          <td>
            Juan undefined
          </td>
          <td>
            GLC300 2023 blue
          </td>
          <td>
            2022-12-10 10:00:00
          </td>
          <td>
            2022-12-20 11:00:00
          </td>
          <td>
            <button
              id="remove__rid-20221210-000001"
              role="button"
            >
              Remove
            </button>
          </td>
        </tr>
        <tr
          id="rid-20221210-000002"
        >
          <td>
            rid-20221210-000002
          </td>
          <td>
            Pepito undefined
          </td>
          <td>
            Q5 2020 white
          </td>
          <td>
            2022-12-10 10:30:00
          </td>
          <td>
            2022-12-20 11:10:00
          </td>
          <td>
            <button
              id="remove__rid-20221210-000002"
              role="button"
            >
              Remove
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
